File: backend/app/infra/outbox.py
"""DomainEvent 同事务 Outbox 的 SQLAlchemy Core 仓储。"""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

from sqlalchemy import insert, or_, select, update
from sqlalchemy.engine import Connection

from app.domain.common import new_object_id
from app.domain.errors import NotFoundError
from app.infra.repositories._common import ensure_project_writable
from app.infra.task2_schema import outbox_messages


@dataclass(frozen=True)
class OutboxMessage:
    """描述待投递或已投递消息，正文只保存事件的 JSON 快照。"""

    id: str
    project_id: str | None
    event_id: str
    topic: str
    payload: dict[str, Any]
    created_at: datetime
    published_at: datetime | None
    status: str
    retry_count: int
    last_error: str | None
    available_at: datetime | None


def _aware(value: datetime | None) -> datetime | None:
    """恢复 SQLite DateTime 的 UTC 时区信息。"""
    if value is None or value.tzinfo is not None:
        return value
    return value.replace(tzinfo=timezone.utc)


class OutboxRepository:
    """在调用方 UnitOfWork 的连接中写入、查询和标记 Outbox 消息。"""

    def enqueue(
        self,
        connection: Connection,
        event: Any,
        *,
        topic: str = "domain.events",
        project_id: str | None = None,
    ) -> OutboxMessage:
        """把已构造 DomainEvent 写入待发布集合，唯一 event_id 防止重复投递记录。"""
        payload = event.model_dump(mode="json", by_alias=False)
        resolved_project_id = project_id if project_id is not None else self._project_id(connection, event)
        self._ensure_project_writable_if_scoped(connection, resolved_project_id)
        values = {
            "id": new_object_id("outbox"),
            "project_id": resolved_project_id,
            "event_id": event.event_id,
            "topic": topic,
            "payload_json": json.dumps(payload, ensure_ascii=False, sort_keys=True, separators=(",", ":"), allow_nan=False),
            "created_at": event.occurred_at,
            "status": "pending",
            "retry_count": 0,
            "available_at": event.occurred_at,
        }
        connection.execute(insert(outbox_messages).values(**values))
        return self._message_from_values(values)

    def list_unpublished(self, connection: Connection, limit: int = 100) -> list[OutboxMessage]:
        """按创建顺序返回尚未发布且已到可用时间的消息。"""
        if limit <= 0:
            raise ValueError("limit must be positive")
        now = datetime.now(timezone.utc)
        rows = connection.execute(
            select(outbox_messages)
            .where(
                outbox_messages.c.published_at.is_(None),
                outbox_messages.c.status != "published",
                or_(outbox_messages.c.available_at.is_(None), outbox_messages.c.available_at <= now),
            )
            .order_by(outbox_messages.c.created_at, outbox_messages.c.id)
            .limit(limit)
        ).mappings().all()
        return [self._message_from_row(row) for row in rows]

    def mark_published(self, connection: Connection, message_id: str, published_at: datetime | None = None) -> OutboxMessage:
        """幂等标记消息已发布；重复标记返回已发布快照，不产生错误。"""
        current = connection.execute(select(outbox_messages).where(outbox_messages.c.id == message_id)).mappings().first()
        if current is None:
            raise NotFoundError(data={"outboxId": message_id})
        self._ensure_project_writable_if_scoped(connection, current["project_id"])
        if current["published_at"] is None or current["status"] != "published":
            connection.execute(
                update(outbox_messages).where(outbox_messages.c.id == message_id).values(
                    published_at=published_at or datetime.now(timezone.utc), status="published"
                )
            )
        row = connection.execute(select(outbox_messages).where(outbox_messages.c.id == message_id)).mappings().one()
        return self._message_from_row(row)

    def mark_failed(
        self,
        connection: Connection,
        message_id: str,
        *,
        error: str,
        available_at: datetime | None = None,
    ) -> OutboxMessage:
        """记录一次投递失败并递增重试次数，失败消息仍保留在 Outbox。"""
        current = self._get_row(connection, message_id)
        self._ensure_project_writable_if_scoped(connection, current["project_id"])
        connection.execute(
            update(outbox_messages)
            .where(outbox_messages.c.id == message_id, outbox_messages.c.published_at.is_(None))
            .values(
                status="failed",
                retry_count=current["retry_count"] + 1,
                last_error=error,
                available_at=available_at or datetime.now(timezone.utc),
            )
        )
        return self._message_from_row(self._get_row(connection, message_id))

    def record_failure(
        self,
        connection: Connection,
        message_id: str,
        *,
        error: str,
        available_at: datetime | None = None,
    ) -> OutboxMessage:
        """提供语义明确的失败记录别名，供投递器按领域命名调用。"""
        return self.mark_failed(connection, message_id, error=error, available_at=available_at)

    def schedule_retry(
        self,
        connection: Connection,
        message_id: str,
        *,
        available_at: datetime,
    ) -> OutboxMessage:
        """把未发布消息重新排入 pending 队列，不清除失败审计信息。"""
        current = self._get_row(connection, message_id)
        self._ensure_project_writable_if_scoped(connection, current["project_id"])
        connection.execute(
            update(outbox_messages)
            .where(outbox_messages.c.id == message_id, outbox_messages.c.published_at.is_(None))
            .values(status="pending", available_at=available_at)
        )
        return self._message_from_row(self._get_row(connection, message_id))

    def retry(
        self,
        connection: Connection,
        message_id: str,
        *,
        available_at: datetime,
    ) -> OutboxMessage:
        """提供简短的重试调度别名，保持失败与重试都使用同一事务连接。"""
        return self.schedule_retry(connection, message_id, available_at=available_at)

    @staticmethod
    def _project_id(connection: Connection, event: Any) -> str | None:
        """读取 DomainEvent 的可选项目范围，不为不存在的测试聚合制造 FK。"""
        value = getattr(event, "project_id", None)
        return value

    @staticmethod
    def _message_from_values(values: dict[str, Any]) -> OutboxMessage:
        """把刚插入的值转换为仓储返回合同。"""
        return OutboxMessage(
            id=values["id"], project_id=values["project_id"], event_id=values["event_id"], topic=values["topic"],
            payload=json.loads(values["payload_json"]), created_at=_aware(values["created_at"]), published_at=None,
            status=values["status"], retry_count=values["retry_count"], last_error=None,
            available_at=_aware(values["available_at"]),
        )

    @staticmethod
    def _message_from_row(row: Any) -> OutboxMessage:
        """把数据库行恢复为不可变 OutboxMessage。"""
        return OutboxMessage(
            id=row["id"], project_id=row["project_id"], event_id=row["event_id"], topic=row["topic"],
            payload=json.loads(row["payload_json"]), created_at=_aware(row["created_at"]),
            published_at=_aware(row["published_at"]), status=row["status"], retry_count=row["retry_count"],
            last_error=row["last_error"], available_at=_aware(row["available_at"]),
        )

    @staticmethod
    def _get_row(connection: Connection, message_id: str) -> Any:
        """读取 Outbox 当前行，统一处理不存在消息的稳定错误。"""
        row = connection.execute(select(outbox_messages).where(outbox_messages.c.id == message_id)).mappings().first()
        if row is None:
            raise NotFoundError(data={"outboxId": message_id})
        return row

    @staticmethod
    def _ensure_project_writable_if_scoped(connection: Connection, project_id: str | None) -> None:
        """仅对带项目范围的消息执行生命周期保护，系统消息保持可用。"""
        if project_id is not None:
            ensure_project_writable(connection, project_id)

File: backend/src/infra/outbox.ts
import BetterSqlite3 from "better-sqlite3";
import { newObjectId, utcNow } from "../domain/common.js";
import { ReadOnlyProjectError } from "../domain/errors.js";
import { jsonText, ensureProjectWritable } from "./repositories/common.js";

/** Outbox 消息的脱敏持久化模型。 */
export type OutboxMessage = {
  id: string;
  projectId: string | null;
  eventId: string;
  topic: string;
  payload: Record<string, unknown>;
  createdAt: string;
  publishedAt: string | null;
  status: string;
  retryCount: number;
  lastError: string | null;
  availableAt: string | null;
};
/** 在同一 SQLite 事务中创建、发布和重试 Outbox 消息。 */
export class OutboxRepository {
  /** 为已提交领域事件创建一条唯一 Outbox 消息。 */
  enqueue(
    connection: BetterSqlite3.Database,
    input: {
      projectId?: string | null;
      eventId: string;
      topic: string;
      payload: Record<string, unknown>;
      availableAt?: string | null;
    },
  ): OutboxMessage {
    if (input.projectId) ensureProjectWritable(connection, input.projectId);
    const message: OutboxMessage = {
      id: newObjectId("outbox"),
      projectId: input.projectId ?? null,
      eventId: input.eventId,
      topic: input.topic,
      payload: input.payload,
      createdAt: utcNow(),
      publishedAt: null,
      status: "pending",
      retryCount: 0,
      lastError: null,
      availableAt: input.availableAt ?? null,
    };
    connection
      .prepare(
        "INSERT INTO outbox_messages (id,project_id,event_id,topic,payload_json,created_at,published_at,status,retry_count,last_error,available_at) VALUES (?,?,?,?,?,?,?,?,?,?,?)",
      )
      .run(
        message.id,
        message.projectId,
        message.eventId,
        message.topic,
        jsonText(message.payload),
        message.createdAt,
        null,
        message.status,
        message.retryCount,
        null,
        message.availableAt,
      );
    return message;
  }
  /** 返回未发布且到达可用时间的消息。 */
  listUnpublished(
    connection: BetterSqlite3.Database,
    limit = 100,
  ): OutboxMessage[] {
    const rows = connection
      .prepare(
        "SELECT * FROM outbox_messages WHERE published_at IS NULL AND status IN ('pending','retry') AND (available_at IS NULL OR available_at<=?) ORDER BY created_at,id LIMIT ?",
      )
      .all(utcNow(), limit) as OutboxRow[];
    return rows.map(fromRow);
  }
  /** 标记消息已发布。 */
  markPublished(
    connection: BetterSqlite3.Database,
    messageId: string,
    publishedAt = utcNow(),
  ): OutboxMessage {
    const row = this.getRow(connection, messageId);
    if (row.project_id) ensureProjectWritable(connection, row.project_id);
    connection
      .prepare("UPDATE outbox_messages SET published_at=?,status=? WHERE id=?")
      .run(publishedAt, "published", messageId);
    return fromRow(this.getRow(connection, messageId));
  }
  /** 记录永久失败。 */
  markFailed(
    connection: BetterSqlite3.Database,
    messageId: string,
    error: string,
  ): OutboxMessage {
    const row = this.getRow(connection, messageId);
    if (row.project_id) ensureProjectWritable(connection, row.project_id);
    connection
      .prepare(
        "UPDATE outbox_messages SET status=?,last_error=?,retry_count=retry_count+1 WHERE id=?",
      )
      .run("failed", error, messageId);
    return fromRow(this.getRow(connection, messageId));
  }
  /** 安排一次可重复消费的重试。 */
  scheduleRetry(
    connection: BetterSqlite3.Database,
    messageId: string,
    availableAt: string,
    error: string,
  ): OutboxMessage {
    const row = this.getRow(connection, messageId);
    if (row.project_id) ensureProjectWritable(connection, row.project_id);
    connection
      .prepare(
        "UPDATE outbox_messages SET status=?,last_error=?,available_at=?,retry_count=retry_count+1 WHERE id=?",
      )
      .run("retry", error, availableAt, messageId);
    return fromRow(this.getRow(connection, messageId));
  }
  /** 别名：将消息重新放回重试队列。 */
  retry(
    connection: BetterSqlite3.Database,
    messageId: string,
    availableAt: string,
    error = "retry requested",
  ): OutboxMessage {
    return this.scheduleRetry(connection, messageId, availableAt, error);
  }
  private getRow(connection: BetterSqlite3.Database, id: string): OutboxRow {
    const row = connection
      .prepare("SELECT * FROM outbox_messages WHERE id=?")
      .get(id) as OutboxRow | undefined;
    if (!row) throw new Error("outbox message not found");
    return row;
  }
}
type OutboxRow = {
  id: string;
  project_id: string | null;
  event_id: string;
  topic: string;
  payload_json: string;
  created_at: string;
  published_at: string | null;
  status: string;
  retry_count: number;
  last_error: string | null;
  available_at: string | null;
};
function fromRow(row: OutboxRow): OutboxMessage {
  return {
    id: row.id,
    projectId: row.project_id,
    eventId: row.event_id,
    topic: row.topic,
    payload: JSON.parse(row.payload_json) as Record<string, unknown>,
    createdAt: row.created_at,
    publishedAt: row.published_at,
    status: row.status,
    retryCount: row.retry_count,
    lastError: row.last_error,
    availableAt: row.available_at,
  };
}
